File: frontend/src/lib/api/documentUpload.ts
/**
 * Document Upload API Client
 * Handles file uploads and document analysis requests
 */

const API_BASE_URL = import.meta.env.VITE_API_URL || 'http://localhost:8000';

export interface DocumentAnalysisResponse {
  file_id: string;
  filename: string;
  file_type: string;
  file_size: number;
  analysis_complete: boolean;
  extracted_text: string;
  key_value_pairs: Array<{ key: string; value: string }>;
  tables: Array<any>;
  entities: Array<any>;
  metadata: Record<string, any>;
}

export interface ImageUploadResponse {
  file_id: string;
  filename: string;
  file_size: number;
  extracted_text: string;
  metadata?: Record<string, any>;
  warning?: string;
  error?: string;
}

export interface AudioUploadResponse {
  file_id: string;
  filename: string;
  file_size: number;
  note: string;
}

/**
 * Get authorization header if user is logged in
 */
function getAuthHeader(): HeadersInit {
  const token = localStorage.getItem('access_token');
  if (token) {
    return {
      'Authorization': `Bearer ${token}`
    };
  }
  return {};
}

/**
 * Upload and analyze a document
 *
 * @param file - The document file to upload
 * @param analyze - Whether to run AI analysis on the document (default: true)
 * @returns Document analysis results
 *
 * @example
 * ```tsx
 * const result = await uploadDocument(file, true);
 * console.log('Extracted text:', result.extracted_text);
 * ```
 */
export async function uploadDocument(
  file: File,
  analyze: boolean = true
): Promise<DocumentAnalysisResponse> {
  const formData = new FormData();
  formData.append('file', file);

  const url = new URL(`${API_BASE_URL}/api/v1/upload/document`);
  if (!analyze) {
    url.searchParams.append('analyze', 'false');
  }

  const response = await fetch(url.toString(), {
    method: 'POST',
    headers: getAuthHeader(),
    body: formData
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || 'Failed to upload document');
  }

  return response.json();
}

/**
 * Upload an image and extract text via OCR
 *
 * @param file - The image file to upload
 * @param extractText - Whether to extract text from the image (default: true)
 * @returns Image upload and OCR results
 *
 * @example
 * ```tsx
 * const result = await uploadImage(imageFile);
 * console.log('OCR text:', result.extracted_text);
 * ```
 */
export async function uploadImage(
  file: File,
  extractText: boolean = true
): Promise<ImageUploadResponse> {
  const formData = new FormData();
  formData.append('file', file);

  const url = new URL(`${API_BASE_URL}/api/v1/upload/image`);
  if (!extractText) {
    url.searchParams.append('extract_text', 'false');
  }

  const response = await fetch(url.toString(), {
    method: 'POST',
    headers: getAuthHeader(),
    body: formData
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || 'Failed to upload image');
  }

  return response.json();
}

/**
 * Upload an audio file
 *
 * @param file - The audio file to upload
 * @returns Audio upload result
 *
 * @example
 * ```tsx
 * const result = await uploadAudio(audioFile);
 * console.log('Audio uploaded:', result.file_id);
 * ```
 */
export async function uploadAudio(
  file: File
): Promise<AudioUploadResponse> {
  const formData = new FormData();
  formData.append('file', file);

  const response = await fetch(`${API_BASE_URL}/api/v1/upload/audio`, {
    method: 'POST',
    headers: getAuthHeader(),
    body: formData
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || 'Failed to upload audio');
  }

  return response.json();
}

/**
 * Delete an uploaded file
 *
 * @param fileId - The ID of the file to delete
 * @returns Deletion confirmation
 *
 * @example
 * ```tsx
 * await deleteUploadedFile('123e4567-e89b-12d3-a456-426614174000');
 * ```
 */
export async function deleteUploadedFile(
  fileId: string
): Promise<{ message: string; file_id: string }> {
  const token = localStorage.getItem('access_token');
  if (!token) {
    throw new Error('Authentication required to delete files');
  }

  const response = await fetch(
    `${API_BASE_URL}/api/v1/upload/file/${fileId}`,
    {
      method: 'DELETE',
      headers: {
        'Authorization': `Bearer ${token}`
      }
    }
  );

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || 'Failed to delete file');
  }

  return response.json();
}

/**
 * Helper function to validate file type
 *
 * @param file - File to validate
 * @param allowedTypes - Array of allowed MIME types or extensions
 * @returns True if file type is allowed
 */
export function validateFileType(
  file: File,
  allowedTypes: string[]
): boolean {
  const fileExtension = '.' + file.name.split('.').pop()?.toLowerCase();
  const mimeType = file.type.toLowerCase();

  return allowedTypes.some(
    (type) =>
      type.toLowerCase() === fileExtension ||
      type.toLowerCase() === mimeType ||
      (type.includes('*') && mimeType.startsWith(type.split('*')[0]))
  );
}

/**
 * Helper function to validate file size
 *
 * @param file - File to validate
 * @param maxSizeMB - Maximum size in megabytes
 * @returns True if file size is within limit
 */
export function validateFileSize(
  file: File,
  maxSizeMB: number = 50
): boolean {
  const maxSizeBytes = maxSizeMB * 1024 * 1024;
  return file.size <= maxSizeBytes;
}

/**
 * Format file size for display
 *
 * @param bytes - File size in bytes
 * @returns Formatted file size string (e.g., "2.5 MB")
 */
export function formatFileSize(bytes: number): string {
  if (bytes === 0) return '0 Bytes';

  const k = 1024;
  const sizes = ['Bytes', 'KB', 'MB', 'GB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));

  return Math.round((bytes / Math.pow(k, i)) * 100) / 100 + ' ' + sizes[i];
}

/**
 * Allowed file types for documents
 */
export const ALLOWED_DOCUMENT_TYPES = [
  '.pdf',
  '.doc',
  '.docx',
  '.txt',
  '.rtf',
  'application/pdf',
  'application/msword',
  'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
  'text/plain'
];

/**
 * Allowed file types for images
 */
export const ALLOWED_IMAGE_TYPES = [
  '.jpg',
  '.jpeg',
  '.png',
  '.bmp',
  '.tiff',
  '.tif',
  'image/jpeg',
  'image/png',
  'image/bmp',
  'image/tiff'
];

/**
 * Allowed file types for audio
 */
export const ALLOWED_AUDIO_TYPES = [
  '.mp3',
  '.wav',
  '.m4a',
  '.ogg',
  '.webm',
  'audio/mpeg',
  'audio/wav',
  'audio/mp4',
  'audio/ogg',
  'audio/webm'
];

File: backend/app/api/document_upload.py
"""
Document Upload and Processing API
Handles file uploads, document analysis, and extraction
"""

import os
import uuid
import logging
from pathlib import Path
from typing import Annotated, Optional
from fastapi import APIRouter, UploadFile, File, Depends, HTTPException, status
from fastapi.responses import JSONResponse
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_db
from app.models.user import User
from app.api.deps import get_current_user, get_current_user_optional
from app.services.document_intelligence import get_document_intelligence_service
from app.services.openrouter_vision import get_openrouter_vision_service
from pydantic import BaseModel

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/upload", tags=["Document Upload"])

# Configuration
UPLOAD_DIR = Path(os.getenv("UPLOAD_DIR", "./uploads"))
UPLOAD_DIR.mkdir(parents=True, exist_ok=True)

MAX_FILE_SIZE = 50 * 1024 * 1024  # 50MB
ALLOWED_EXTENSIONS = {
    # Documents
    ".pdf", ".doc", ".docx", ".txt", ".rtf",
    # Images
    ".jpg", ".jpeg", ".png", ".bmp", ".tiff", ".tif",
    # Other
    ".eml", ".msg"
}


class DocumentAnalysisResponse(BaseModel):
    """Response model for document analysis"""
    file_id: str
    filename: str
    file_type: str
    file_size: int
    analysis_complete: bool
    extracted_text: str
    key_value_pairs: list
    tables: list
    entities: list
    metadata: dict


@router.post("/document", response_model=DocumentAnalysisResponse)
async def upload_and_analyze_document(
    file: UploadFile = File(...),
    analyze: bool = True,
    current_user: Annotated[User | None, Depends(get_current_user_optional)] = None,
    db: Annotated[AsyncSession, Depends(get_db)] = None
):
    """
    Upload a document and optionally analyze it with AI

    Args:
        file: The uploaded file
        analyze: Whether to run AI analysis on the document
        current_user: Optional authenticated user
        db: Database session

    Returns:
        Document analysis results
    """
    try:
        # Validate file
        if not file.filename:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="No filename provided"
            )

        # Check file extension
        file_ext = Path(file.filename).suffix.lower()
        if file_ext not in ALLOWED_EXTENSIONS:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"File type {file_ext} not allowed. Allowed types: {', '.join(ALLOWED_EXTENSIONS)}"
            )

        # Read and validate file size
        contents = await file.read()
        if len(contents) > MAX_FILE_SIZE:
            raise HTTPException(
                status_code=status.HTTP_413_REQUEST_ENTITY_TOO_LARGE,
                detail=f"File size exceeds maximum allowed size of {MAX_FILE_SIZE / 1024 / 1024}MB"
            )

        # Generate unique file ID and save file
        file_id = str(uuid.uuid4())
        safe_filename = f"{file_id}{file_ext}"
        file_path = UPLOAD_DIR / safe_filename

        with open(file_path, "wb") as f:
            f.write(contents)

        logger.info(f"File uploaded successfully: {file.filename} -> {safe_filename}")

        # Initialize response
        response = {
            "file_id": file_id,
            "filename": file.filename,
            "file_type": file_ext,
            "file_size": len(contents),
            "analysis_complete": False,
            "extracted_text": "",
            "key_value_pairs": [],
            "tables": [],
            "entities": [],
            "metadata": {}
        }

        # Perform AI analysis if requested
        if analyze:
            analysis_result = None
            service_used = None

            # Try OpenRouter first for images (best for vision tasks)
            if file_ext in [".jpg", ".jpeg", ".png", ".bmp", ".tiff", ".tif", ".gif", ".webp"]:
                try:
                    openrouter_service = get_openrouter_vision_service()

                    if "invoice" in file.filename.lower() or "rechnung" in file.filename.lower():
                        analysis_result = await openrouter_service.analyze_invoice(str(file_path))
                    else:
                        analysis_result = await openrouter_service.analyze_document_image(str(file_path))

                    service_used = "openrouter"
                    logger.info(f"Image analyzed with OpenRouter: {file.filename}")
                except ValueError as e:
                    logger.warning(f"OpenRouter not available: {str(e)}")
                except Exception as e:
                    logger.error(f"OpenRouter analysis failed: {str(e)}")

            # Fallback to Azure Document Intelligence if OpenRouter failed or for PDFs
            if not analysis_result:
                try:
                    doc_intel_service = get_document_intelligence_service()

                    # Determine analysis type based on file extension
                    if file_ext in [".jpg", ".jpeg", ".png", ".bmp", ".tiff", ".tif"]:
                        analysis_result = await doc_intel_service.analyze_image(str(file_path))
                    elif "invoice" in file.filename.lower() or "rechnung" in file.filename.lower():
                        analysis_result = await doc_intel_service.analyze_invoice(str(file_path))
                    elif "receipt" in file.filename.lower() or "quittung" in file.filename.lower():
                        analysis_result = await doc_intel_service.analyze_receipt(str(file_path))
                    else:
                        analysis_result = await doc_intel_service.analyze_legal_document(str(file_path))

                    service_used = "azure"
                    logger.info(f"Document analyzed with Azure: {file.filename}")
                except ValueError as e:
                    logger.warning(f"Azure Document Intelligence not available: {str(e)}")
                except Exception as e:
                    logger.error(f"Azure analysis failed: {str(e)}")

            # Process results if we got any
            if analysis_result:
                # Update response with analysis results
                response.update({
                    "analysis_complete": True,
                    "extracted_text": analysis_result.get("text", "") or analysis_result.get("full_text", "") or analysis_result.get("extracted_text", ""),
                    "key_value_pairs": analysis_result.get("key_value_pairs", []) or analysis_result.get("key_fields", []),
                    "tables": analysis_result.get("tables", []),
                    "entities": analysis_result.get("entities", []),
                    "metadata": {
                        **analysis_result.get("metadata", {}),
                        "service_used": service_used
                    }
                })

                logger.info(f"Document analysis completed for {file.filename} using {service_used}")
            else:
                # No analysis service available
                response["metadata"]["warning"] = "AI analysis not available - no service configured (set OPENROUTER_API_KEY or DOCINTEL_API_KEY)"

        return response

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error uploading document: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error uploading document: {str(e)}"
        )


@router.post("/image")
async def upload_image(
    file: UploadFile = File(...),
    extract_text: bool = True,
    current_user: Annotated[User | None, Depends(get_current_user_optional)] = None
):
    """
    Upload an image and extract text via OCR

    Args:
        file: The uploaded image file
        extract_text: Whether to extract text from the image
        current_user: Optional authenticated user

    Returns:
        Image upload result with OCR text
    """
    try:
        # Validate image file
        file_ext = Path(file.filename).suffix.lower()
        if file_ext not in [".jpg", ".jpeg", ".png", ".bmp", ".tiff", ".tif"]:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid image format. Allowed: JPG, PNG, BMP, TIFF"
            )

        # Read file
        contents = await file.read()
        if len(contents) > MAX_FILE_SIZE:
            raise HTTPException(
                status_code=status.HTTP_413_REQUEST_ENTITY_TOO_LARGE,
                detail=f"File size exceeds {MAX_FILE_SIZE / 1024 / 1024}MB"
            )

        # Save file
        file_id = str(uuid.uuid4())
        safe_filename = f"{file_id}{file_ext}"
        file_path = UPLOAD_DIR / safe_filename

        with open(file_path, "wb") as f:
            f.write(contents)

        result = {
            "file_id": file_id,
            "filename": file.filename,
            "file_size": len(contents),
            "extracted_text": ""
        }

        # Extract text if requested
        if extract_text:
            try:
                doc_intel_service = get_document_intelligence_service()
                analysis = await doc_intel_service.analyze_image(str(file_path))
                result["extracted_text"] = analysis.get("text", "")
                result["metadata"] = analysis.get("metadata", {})
            except ValueError:
                result["warning"] = "OCR not available - service not configured"
            except Exception as e:
                logger.error(f"OCR failed for {file.filename}: {str(e)}")
                result["error"] = str(e)

        return result

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error uploading image: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )


@router.post("/audio")
async def upload_audio(
    file: UploadFile = File(...),
    current_user: Annotated[User | None, Depends(get_current_user_optional)] = None
):
    """
    Upload an audio file (for future speech-to-text processing)

    Args:
        file: The uploaded audio file
        current_user: Optional authenticated user

    Returns:
        Audio upload result
    """
    try:
        # Validate audio file
        file_ext = Path(file.filename).suffix.lower()
        if file_ext not in [".mp3", ".wav", ".m4a", ".ogg", ".webm"]:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid audio format. Allowed: MP3, WAV, M4A, OGG, WEBM"
            )

        # Read file
        contents = await file.read()
        if len(contents) > MAX_FILE_SIZE:
            raise HTTPException(
                status_code=status.HTTP_413_REQUEST_ENTITY_TOO_LARGE,
                detail=f"File size exceeds {MAX_FILE_SIZE / 1024 / 1024}MB"
            )

        # Save file
        file_id = str(uuid.uuid4())
        safe_filename = f"{file_id}{file_ext}"
        file_path = UPLOAD_DIR / safe_filename

        with open(file_path, "wb") as f:
            f.write(contents)

        return {
            "file_id": file_id,
            "filename": file.filename,
            "file_size": len(contents),
            "note": "Audio file uploaded successfully. Speech-to-text processing will be added in future updates."
        }

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error uploading audio: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )


@router.delete("/file/{file_id}")
async def delete_uploaded_file(
    file_id: str,
    current_user: Annotated[User, Depends(get_current_user)]
):
    """
    Delete an uploaded file

    Args:
        file_id: The file ID to delete
        current_user: Authenticated user

    Returns:
        Deletion confirmation
    """
    try:
        # Find file with this ID
        found = False
        for file_path in UPLOAD_DIR.glob(f"{file_id}*"):
            file_path.unlink()
            found = True
            logger.info(f"Deleted file: {file_path}")

        if not found:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="File not found"
            )

        return {"message": "File deleted successfully", "file_id": file_id}

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error deleting file {file_id}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )
